File: algorithms/algorithms-cplusplus/leetcode/Question_0971_Flip_Binary_Tree_To_Match_Preorder_Traversal.cc
#include <vector>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
 public:
  vector<int> flipMatchVoyage(TreeNode* root, vector<int>& voyage) {
    int i = 0;
    vector<int> res;
    if (!help(root, voyage, i, res) || i != voyage.size()) {
      return {-1};
    }
    return res;
  }

 private:
  bool help(TreeNode* root, vector<int>& voyage, int& i, vector<int>& res) {
    if (!root) {
      return true;
    }
    if (root->val != voyage[i++]) {
      return false;
    }
    if (root->left && root->left->val != voyage[i]) {
      res.push_back(root->val);
      return help(root->right, voyage, i, res) && help(root->left, voyage, i, res);
    }
    return help(root->left, voyage, i, res) && help(root->right, voyage, i, res);
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0971_Flip_Binary_Tree_To_Match_Preorder_Traversal.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Question_0971_Flip_Binary_Tree_To_Match_Preorder_Traversal {
    // Definition for a binary tree node.
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    private int i = 0;
    private List<Integer> res = new ArrayList<>();

    public List<Integer> flipMatchVoyage(TreeNode root, int[] voyage) {
        return dfs(root, voyage) ? res : Arrays.asList(-1);
    }

    public Boolean dfs(TreeNode node, int[] v) {
        if (node == null) {
            return true;
        }
        if (node.val != v[i++]) {
            return false;
        }
        if (node.left != null && node.left.val != v[i]) {
            res.add(node.val);
            return dfs(node.right, v) && dfs(node.left, v);
        }
        return dfs(node.left, v) && dfs(node.right, v);
    }
}
